DOC Document the file usage table

emteknetnz opened the pull request:
Issue: https://github.com/silverstripe/silverstripe-admin/issues/1098

Blocked by https://github.com/silverstripe/silverstripe-admin/pull/1139

maxime-rainville commented on docs/en/02_Developer_Guides/14_Files/07_File_Usage.md:
```diff
@@ -0,0 +1,63 @@
+---
+
+# File Usage
+
```
Maybe add a bit of an intro about what the feature does.

```suggestion

CMS users can view where a file is used by accessing the "Used On" tab in the file editor. This feature allows them to identify what DataObjects depend on the File. 

```

emteknetnz replied:
Added

maxime-rainville commented on docs/en/02_Developer_Guides/14_Files/07_File_Usage.md:
```diff
@@ -0,0 +1,63 @@
+In the Files section of the CMS, click on a file to see file details.  Within the file details panel there is a "Used on" tab that shows a table of Pages and other DataObjects where the file is used throughout the website.
+
+## Customising the File "Used on" table in the Files section (asset-admin)
+
```
Maybe add a blurb to contextualise this a bit.

```suggestion

Your project specific DataObject will automatically be displayed on the "Used on". This may not always be desirable, especially when working with background DataObjects the user can not interact with directly. Extensions can be applied to the `UsedOnTable` class to suppress specific entries.

```

emteknetnz replied:
Added

Cheddam reviewed (changes requested):
It feels like this could use some fleshing out - not having been across this API's development, I don't get a clear understanding of how I should be interacting with it from reading this doc.

- Would it make more sense to break out each of the available extension points and fully describe its purpose, behaviour, and any potential pitfalls? Performance implications seem particular important - `updateUsageDataObject()` in looks like it operates on every single DataObject in the table, which could get heavy very easily. Some of the methods could probably benefit from multiple example implementations, too.
- Are there any other sections of the documentation that could include links to this doc to make it more discoverable?
- Please reformat to [a max of 120 characters per line](https://docs.silverstripe.org/en/4/contributing/documentation/#structure)
- An intro to this API needs to be included in the 4.7.0 / 4.7.0-beta1 changelog. In particular it'd be good to outline what change in behaviour developers will see from the Used On table when upgrading.

emteknetnz commented:
@Cheddam 
- I've added inline comments to the the extension hook examples to further elaborate what each of the extension hooks do
- I've reformatted to 120 chars
- I've added an intro in the beta changelog

I don't know where else in the docs it would make sense to link to this.  It will have its own section in https://docs.silverstripe.org/en/4/developer_guides/files/#files which seems like it's good enough?

brynwhyman commented on docs/en/02_Developer_Guides/14_Files/07_File_Usage.md:
```diff
@@ -0,0 +1,84 @@
+
+# File Usage
+
+CMS users can view where a file is used by accessing the "Used On" tab in the file editor. This feature allows them to identify what DataObjects depend on the File.
```
```suggestion
CMS users can view where a file is used by accessing the Used on tab in the Files section. This feature allows them to identify what DataObjects depend on the file.
```

brynwhyman replied:
These suggestions are more in line with the writing structure here: https://userhelp.silverstripe.org/en/4/creating_pages_and_content/creating_and_editing_content/images_and_documents/#file-usage

brynwhyman commented on docs/en/02_Developer_Guides/14_Files/07_File_Usage.md:
```diff
@@ -0,0 +1,84 @@
+CMS users can view where a file is used by accessing the "Used On" tab in the file editor. This feature allows them to identify what DataObjects depend on the File.
+
+In the Files section of the CMS, click on a file to see file details.  Within the file details panel there is a
+"Used on" tab that shows a table of Pages and other DataObjects where the file is used throughout the website.
```
```suggestion
Used on tab that shows a table of Pages and other DataObjects where the file is used throughout the website.
```

brynwhyman commented on docs/en/02_Developer_Guides/14_Files/07_File_Usage.md:
```diff
@@ -0,0 +1,84 @@
+
+## Customising the File "Used on" table in the Files section (asset-admin)
+
+Your project specific DataObject will automatically be displayed on the "Used on". This may not always be desirable, especially when working with background DataObjects the user can not interact with directly. Extensions can be applied
```
```suggestion
Your project specific DataObject will automatically be displayed on the Used on tab. This may not always be desirable, especially when working with background DataObjects the user can not interact with directly. Extensions can be applied
```

brynwhyman commented on docs/en/02_Developer_Guides/14_Files/07_File_Usage.md:
```diff
@@ -0,0 +1,84 @@
+class UsedOnTableExtension extends Extension
+{
+
+    // This extension hook will prevent type(s) of DataObjects from showing on the used on table
```
```suggestion
    // This extension hook will prevent type(s) of DataObjects from showing on the Used on tab in the Files section
```

brynwhyman commented on docs/en/04_Changelogs/beta/4.7.0-beta1.md:
```diff
@@ -10,6 +10,19 @@
 
+### Updated file usage table
+
+CMS users can view where a file is used by accessing the "Used On" tab in the file editor. This feature allows them to identify what DataObjects depend on the File.
```
```suggestion
CMS users can view where a file is used by accessing the Used on tab in the Files section. This feature allows them to identify what DataObjects depend on the file.
```

brynwhyman commented on docs/en/04_Changelogs/beta/4.7.0-beta1.md:
```diff
@@ -10,6 +10,19 @@
+
+CMS users can view where a file is used by accessing the "Used On" tab in the file editor. This feature allows them to identify what DataObjects depend on the File.
+
+The following changes have been made to the used on table:
```
```suggestion
The following changes have been made to the Used on table:
```

brynwhyman commented on docs/en/04_Changelogs/beta/4.7.0-beta1.md:
```diff
@@ -10,6 +10,19 @@
+- Support for linking ancestor relationships in the same table row, for instance showing the Page a Content Block is on
+- Ability to click on a row and navigate to the most relevant `CMSEditLink()`
+- Removal of draft badges
+- Styling tweaks such as reduced padding
+
```
```suggestion
- Styling tweaks such as reduced padding

Your project specific DataObjects will automatically be displayed on this Used on tab. This may not always be desirable, especially when working with background DataObjects the user can not interact with directly. It is recommended you view the CMS Developer Guide for more information on this change, see the [File Usage section](docs/en/02_Developer_Guides/14_Files/07_File_Usage).
```

brynwhyman replied:
Note I haven't tested that link. Dependant on this PR....

emteknetnz replied:
Updated

emteknetnz commented:
@brynwhyman have implemented your feedback and add you as a co-author to the commit message

brynwhyman commented on docs/en/04_Changelogs/beta/4.7.0-beta1.md:
```diff
@@ -10,6 +10,21 @@
+- Data is now sourced from has_one and many_many relations instead of the ownership API
+- This means more data relationships should be shown than before
+- Uses the new RelatedDataService API in the framework module which can also be used for other non-file use cases
+- Support for linking ancestor relationships in the same table row, for instance showing the Page a Content Block is on
```
```suggestion
- Support for linking ancestor relationships in the same table row, for instance showing the Page a Content Block is on
- With expected performance impacts of the table referencing a wider range of DataObject relations there has been a change to lazy-load the table when the Used on tab is selected
```

brynwhyman replied:
One more suggestion @emteknetnz

Cheddam replied:
Added a variation of this 👍

Cheddam commented:
@emteknetnz Two more minor things - the changelog item needs to be added to the table of contents at the top of the doc, and you should replicate this item in the (provisional) 4.7.0 stable changelog for consistency.

In terms of where else in the docs this could be linked from, a note about it in the relationships article might make sense, since some of the examples there use `File`s:
https://docs.silverstripe.org/en/4/developer_guides/model/relations/
